### PR TITLE
Document GroupedDataFrame consistency check

### PR DESCRIPTION
**Pull request by bkamins**
Following the discussion in https://github.com/JuliaData/DataFrames.jl/pull/2809 this PR removes the `@assert` checking consistency of `GroupedDataFrame`.

### CODE REVIEW

**Comment by bkamins** [comment]
As an afterthought - maybe instead of removing the assertion we should change it to:
```
@assert length(getfield(gd, :groups)) <= nrow(getfield(gd, :parent)) corrupt_msg(gd)
```
so that appending to the parent does not trigger the assertion, but removing them does.
@raphaelsaavedra + @nickrobinson251: what do you think (I do not see a scenario when dropping rows might be useful).

**Comment by nickrobinson251** [comment]
Thanks!

> so that appending to the parent does not trigger the assertion, but removing them does

I don't think i have a strong opinion on this add/remove difference in particular, although I _think_ i'd find it surprising to discover adding rows allowed but removing them not.

**Comment by bkamins** [comment]
> i'd find it surprising to discover adding rows allowed but removing them not.

The reason is that we use `@inbounds` in the code. When you ADD rows you will stay within bounds. When you remove rows you will be out of bounds, ans with `@inbounds` it might lead to crashing Julia.

**Comment by nickrobinson251** [comment]
Oh, yes I think it is better to remove the assertion.
Whilst I agree it's not a particularly safe pattern, I don't think it should be an error condition (not least because it's useful, and apparently relied upon by some internal code at Invenia).

Thanks for being so fast to respond on this!

**Comment by bkamins** [comment]
> and apparently relied upon by some internal code at Invenia

OK - then I will merge this PR and make a release after I get an approval. I will just push one more commit to this PR improving a documentation of `GroupedDataFrame`.

**Comment by pdeffebach** [comment]
I want to chime in and say that I don't think this should be merged. 

I think the docs stating "Modifying the underlying data frame may corrupt a grouped data frame" implies that we don't support grouped data frames that become out of sync with their parent data frame. Given this implication, we can throw an error when they get out of sync. Appending to a data frame past its grouping indices should be considered undefined behavior.

**Comment by raphaelsaavedra** [comment]
Just to be clear, I think it would be pretty simple to rewrite our code not to rely on this particular interaction, so I don't think that should be the deciding factor. I commented on the PR because I thought it might have been an unintentional effect of the assertion and possibly an unintentional breaking change. You guys are certainly more qualified to decide on the matters of this package, so feel free to choose what you prefer :) and thanks for the quick response!

**Comment by bkamins** [comment]
Thank you all for the comments. We have the following tension here:

On one side we do not strictly need the `@assert` - all will work without it and I can understand that advanced users, like Invenia, might be relying on undefined behavior to e.g. improve performance of operations.

On the other side the reason why the `@assert` was added was to safeguard the users, mostly newcomers, from unintentionally mutating the parent without learning that the `GroupedDataFrame` got corrupted.

So we have flexibility vs safety tension here. Let us wait to hear what @nalimilan thinks. Maybe @oxinabox would also have an opinion here.

Just to summarize what I normally would recommend (independent on what we decide to do in this PR). Instead of:
```
julia> for g in groupby(df, :y)
           nrow(g) == 1 && append!(df, DataFrame(x=[99], y=first(g.y)))
       end
```
One should write (simplest):
```
julia> for g in groupby(copy(df), :y)
           nrow(g) == 1 && append!(df, DataFrame(x=[99], y=first(g.y)))
       end
```
or (more efficient)
```
julia> aux_df = DataFrame()
julia> for g in groupby(df, :y)
           nrow(g) == 1 && append!(aux_df, DataFrame(x=[99], y=first(g.y)))
       end
julia> append!(df, aux_df) # after the loop is done append the rows
```
Both patterns are strictly preferred to just doing an update of the parent within the loop.

**Comment by nalimilan** [comment]
> As an afterthought - maybe instead of removing the assertion we should change it to:
> 
> ```
> @assert length(getfield(gd, :groups)) <= nrow(getfield(gd, :parent)) corrupt_msg(gd)
> ```
> 
> so that appending to the parent does not trigger the assertion, but removing them does.

Maybe we could adopt a hybrid solution: use the above `@assert` since that can prevent invalid memory accesses (and therefore a crash), but only warn when rows have been added? That would be less disruptive for users, at least as a temporary solution.

Regarding the alternative patterns, is this approach OK?
```julia
df = DataFrame(x = [1, 2, 3], y = [1, 2, 2])
sdf = view(df, 1:nrow(df), :)
for g in groupby(sdf, :y)
    nrow(g) == 1 && append!(df, DataFrame(x=[99], y=first(g.y)))
end
```

**Comment by oxinabox** [comment]
>  but only warn when rows have been added? 

Warnings would be very disruptive.
Because if processing 100 files with this pattern you get spammed 100 times.
Sure I could write a LoggingExtras filter to mute that but easier maybe to rewrite to avoid this pattern.
So it might as well be an error.

**Comment by bkamins** [comment]
> sdf = view(df, 1:nrow(df), :)
for g in groupby(sdf, :y)

Yes this approach is OK and efficient. Even `groupby(view(df, :, :), :y)` is OK and it is easier to write, as `:` does not get any special treatment for row selection (as opposed to column selection). The reason is that now we have:
```
julia> df = DataFrame(a=[1,2,3]);

julia> sdf = view(df, :, :);

julia> getfield(sdf, :rows)
Base.OneTo(3)
```
so as you can see such `sdf` almost does not allocate and will be very fast to index (even it is possible that compiler might optimize out the `Base.OneTo` part - though I have not checked it, but potentially it should do it).

Given the proposal by @nalimilan I would keep the error and recommend the `view` approach. The good thing is that as we are throwing an error now it will be easy to catch and fix the undefined behavior. Just replace any `df` in `groupby(df, cols)` with `view(df, :, :)`.

Let us then wait a few days - if there will be no more comments on this approach I will rewrite this PR so that we keep the `@assert` part and only improve the `GroupedDataFrame` docstring.

**Comment by bkamins** [comment]
Ah - and I would like to thank everyone for reporting the issue and the discussion. I think the solution we have reached for this use case is quite good and have a better understanding of the design intentions and available options in general.

**Comment by bkamins** [comment]
@raphaelsaavedra - are descriptions I have added to the documentation OK?

**Comment by bkamins** [comment]
@mortenpi - could you please kindly have a look at the doctest error? I do not understand why `AssertionError` causes an error in doctests (I would expect it should be handled as described in https://juliadocs.github.io/Documenter.jl/stable/man/doctests/#Exceptions). Thank you!

**Comment by raphaelsaavedra** [comment]
The additions to the documentation are great and super clear, thanks a lot! I have already fixed our internal code by using the suggested method.

File: docs/src/man/split_apply_combine.md
Comment (nalimilan) [inline docs/src/man/split_apply_combine.md:532]:
```suggestion
Note that `GroupedDataFrame` is a view: therefore 
grouping columns of its parent data frame must not be mutated, and
rows must not be added nor removed from it. If the number or rows
of the parent changes then an error is thrown when a child `GroupedDataFrame`
is used:
```

File: docs/src/man/split_apply_combine.md
Comment (nalimilan) [inline docs/src/man/split_apply_combine.md:570]:
```suggestion
Sometimes it is useful to append rows to the source data frame of a
```

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:28]:
```suggestion
After its creation, a `GroupedDataFrame` reflects the grouping of rows that was
valid at the its creation time. Therefore making changes to
```

**Comment by mortenpi** [comment]
> @mortenpi - could you please kindly have a look at the doctest error? I do not understand why `AssertionError` causes an error in doctests (I would expect it should be handled as described in https://juliadocs.github.io/Documenter.jl/stable/man/doctests/#Exceptions). Thank you!

Hmm, I think it's because Documenter only handles errors thrown when executing the code, but not when `show`ing the resulting object (the actual error happens here https://github.com/JuliaDocs/Documenter.jl/blob/v0.27.3/src/DocTests.jl#L319). https://github.com/JuliaDocs/Documenter.jl/issues/1631. But for the time being, I would suggest turning that code snippet into a non-executing `julia-repl` block.

**Comment by bkamins** [comment]
Thank you. I will then force an error by some other operation :)

File: src/groupeddataframe/groupeddataframe.jl
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:29]:
```suggestion
valid at the its creation time. Therefore
grouping columns of its parent data frame must not be mutated, and
rows must not be added nor removed from it.
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**
